### PR TITLE
Keep {} in Markdown rendering

### LINKED ISSUE #10928: Markdown rendering should keep {}
JabRef has internally a field marker that marks a field as `Markdown`. `org.jabref.model.entry.field.FieldProperty#MARKDOWN`

When that field is rendered in the preview, `{}` should be kept.

Example content of a Comment:

````markdown
# test

{ test }

```javascript
function hero() {
}
```
````

307950652-09a3c1b5-dcb5-4553-bce6-766ca5775b84

307951490-65d6369a-f775-4183-89d0-371cf46cc9c1

---

This issue is more about **code reading** and understanding JabRef's archictecture. The "real" fix should be a few lines of code: Checking for the field property `MARKDOWN` and passing the field value with plain latex (and not latex free).

### PR DESCRIPTION
**Pull request by JEANPRBT**
This PR modifies the formatter `HTMLChars` used alongside the formatter `Markdown` to avoid the useless swallow of chars `{` and `}`. It adapts tests accordingly and adds `RemoveBrackets` formatter in relevant `*.layout` files to maintain previous behavior for non-Markdown fields. 
See this [comment](https://github.com/JabRef/jabref/issues/10928#issuecomment-1970094191).

- [x] Adapt tests
- [x] Adapt `*.layout` files

Closes #10928 

Now, Markdown rendering keeps braces, see following screenshot.

braces_kept

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by JEANPRBT** [comment]
@koppor, in your [comment](https://github.com/JabRef/jabref/issues/10928#issuecomment-1970094191) you say the following.

> Thinking more, the LaTeX command handling in HTMLChars should be removed. In the `.layout` files, `LaTeXToUnicode` should be used. - I think, `HTMLChars` will be much smaller then. HTML5 default encoding is unicode, thus `LaTeXToUnicode` is the right formatter.

Do you want me to also cover this in this current PR, or does this deserve another one ?

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 1]
```diff
@@ -51,16 +50,13 @@ void basicFormat(String expected, String input) {
     private static Stream<Arguments> provideLaTeXHighlightingFormattingData() {
         return Stream.of(
                 Arguments.of("<em>hallo</em>", "\\emph{hallo}"),
-                Arguments.of("<em>hallo</em>", "{\\emph hallo}"),
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:54]: Please keep that line. Add {} to the expected output. (See above for reasoning)

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 2]
```diff
@@ -81,7 +77,7 @@ private static Stream<Arguments> provideEquationTestData() {
         return Stream.of(
                 Arguments.of("&dollar;", "\\$"),
                 Arguments.of("&sigma;", "$\\sigma$"),
-                Arguments.of("A 32&nbsp;mA &Sigma;&Delta;-modulator", "A 32~{mA} {$\\Sigma\\Delta$}-modulator")
+                Arguments.of("A 32&nbsp;mA &Sigma;&Delta;-modulator", "A 32~mA $\\Sigma\\Delta$-modulator")
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:80]: The code change is OK. Reason: `{` is not used as parameter for a LaTeX command.

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 3]
```diff
@@ -110,7 +106,7 @@ void newLine(String expected, String input) {
 
     @Test
     void quoteSingle() {
-        assertEquals("&#39;", layout.format("{\\textquotesingle}"));
+        assertEquals("&#39;", layout.format("\\textquotesingle"));
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:109]: The code change is OK. Reason: `{` is not used as parameter for a LaTeX command.

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 4]
```diff
@@ -27,16 +27,15 @@ private static Stream<Arguments> provideBasicFormattingData() {
...
                 Arguments.of("%%%", "\\%\\%\\%"),
-                Arguments.of("People remember 10%, 20%…Oh Really?", "{{People remember 10\\%, 20\\%…Oh Really?}}"),
                 Arguments.of("h&aacute;llo", "h\\'allo"),
                 Arguments.of("&imath; &imath;", "\\i \\i"),
                 Arguments.of("&imath;", "\\i"),
                 Arguments.of("&imath;", "\\{i}"),
                 Arguments.of("&imath;&imath;", "\\i\\i"),
-                Arguments.of("&auml;", "{\\\"{a}}"),
-                Arguments.of("&auml;", "{\\\"a}"),
+                Arguments.of("&auml;", "\\\"{a}"),
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:35]: The old cases should be kept. Add the braces to the expected output.

Reason: modifying tests is OK. However, existing inputs should be kept - and not just removed. (If they are removed, they should be removed with reason!)

Thus, re-add the old test cases.

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 5]
```diff
@@ -27,16 +27,15 @@ private static Stream<Arguments> provideBasicFormattingData() {
                 Arguments.of("hallo", "hallo"),
                 Arguments.of("Réflexions sur le timing de la quantité", "Réflexions sur le timing de la quantité"),
                 Arguments.of("%%%", "\\%\\%\\%"),
-                Arguments.of("People remember 10%, 20%…Oh Really?", "{{People remember 10\\%, 20\\%…Oh Really?}}"),
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:30]: This test should be adapted, because it checks escaped percent signs. Maybe, just add the enclosing double braces to the excpectged output?

**Review by koppor — CHANGES_REQUESTED** [review]
Small comments regarding the test

**Comment by koppor** [comment]
> The only drawback is that it seems to add more newlines than necessary,

Without going deep into the code, I cannot follow. Can you add test cases to `org.jabref.logic.layout.format.MarkdownFormatterTest` so that the current behavior is shown? Then, the new behavior can also be shown in test cases.

Remember that modern Java allows to use `"""` for multiline string constants (including newlines) which can come handy here.

**Comment by JEANPRBT** [comment]
> Small comments regarding the test

This is fixed now. 
> * [x]   Please also add a test case to `org.jabref.logic.layout.format.MarkdownFormatterTest` so that it is really seen that this works.

I should have thought about this before, my bad.

> Since you went deep into fixing HTMLChars, that update can be done in a next PR. It also more about experimentation. --> Test cases in org.jabref.logic.layout.format.MarkdownFormatterTest are needed to ensure that the functionality is unchanged after rewriting HTMLChars.

Ok ! I'll do this in a separate PR then. You're right it will be experimentation, since I don't really see the border between `HTMLChars` and `LaTeXToUnicode` formatters at the moment. I added many test cases to `org.jabref.logic.layout.format.MarkdownFormatterTest` to ensure functionnality will be unchanged in next PR.

**Comment by JEANPRBT** [comment]
> Without going deep into the code, I cannot follow. Can you add test cases to `org.jabref.logic.layout.format.MarkdownFormatterTest` so that the current behavior is shown? Then, the new behavior can also be shown in test cases.

Current behavior is shown using the following test. 
```java
 @Test
    void formatWhenFormattingComplexMarkupThenReturnsOnlyOneLine() {
        assertFalse(markdownFormatter.format("Markup\n\n* list item one\n* list item 2\n\n rest").contains("\n"));
    }
```
The formatter only removes every new line.

**Comment by koppor** [comment]
> Ok ! I'll do this in a separate PR then. You're right it will be experimentation, since I don't really see the border between `HTMLChars` and `LaTeXToUnicode` formatters at the moment.

Yeah, you are right. - I first thought LaTeXToUnicode is executed first and then some Unicode-to-HTML conversion. italics unicode back to `<em>`. However, this might be too complicated and the current approach is OK - even though LaTeX parsing is doubled. -- Maybe, it would make sense to use SnuggleTeX to parse LaTeX and do the conversion based on the DOM tree? (See org.jabref.logic.integrity.LatexIntegrityChecker for some code ideas).

However, LaTeX has a stable syntax, maybe, we leave everything as is.

**Comment by koppor** [comment]
> Current behavior is shown using the following test.

Not quote. Please use `assertEquals`. Left: The new string, right: the original stirng.

USe

```java
  String input = """
     test
  
     some more
     """
```

This is more readable for the human reader than reading `\n`.

**Comment by koppor** [comment]
> Now I have to adapt every `*.layout` files, adding the formatter `RemoveBrackets` for non-Markdown fields. Will be done in a couple of days max.

This is a nightmare for backward compatiblity :)

Thinking aloud:

- When using latex2unicode, these `{}` are removed. Therefore maybe the idea to use that in htmlchars. However, not wanted for markdown
- The htmlchars converter could be parameterized to keep braces. Depending on the mode, it could behave differently. (enum BraceHandling: KEEP, REMOVE)
- maybe the previewer can parameterize the htmlchars with KEEP - and the exporter using layout parameteirzed using REMOVE

OK, the default preview of JabRef needs to be adapted maybe... The normal latex2unicode should be able to handle most cases, thus I think, there probably is no adaption necessary?

**Comment by JEANPRBT** [comment]
> Yeah, you are right. - I first thought LaTeXToUnicode is executed first and then some Unicode-to-HTML conversion. italics unicode back to `\<em\>`. However, this might be too complicated and the current approach is OK - even though LaTeX parsing is doubled. -- Maybe, it would make sense to use SnuggleTeX to parse LaTeX and do the conversion based on the DOM tree? (See org.jabref.logic.integrity.LatexIntegrityChecker for some code ideas).
> 
> However, LaTeX has a stable syntax, maybe, we leave everything as is.

I agree with you, since the current syntax is working properly and changes are not simply LaTeXToUnicode and then Unicode-to-HTML, we can leave everything as is. In any case, perhaps the thymeleaf template engine will fill all these gaps this summer ? ;)
 
> Not quote. Please use `assertEquals`. Left: The new string, right: the original string.
> 
> Use
> 
> ```java
>   String input = """
>      test
>   
>      some more
>      """
> ```
> 
> This is more readable for the human reader than reading `\n`.

I'll implement a couple of tests with quotes then.

**Comment by JEANPRBT** [comment]
> This is a nightmare for backward compatiblity :)
> 
> Thinking aloud:
> 
> * When using latex2unicode, these `{}` are removed. Therefore maybe the idea to use that in htmlchars. However, not wanted for markdown
> * The htmlchars converter could be parameterized to keep braces. Depending on the mode, it could behave differently. (enum BraceHandling: KEEP, REMOVE)
> * maybe the previewer can parameterize the htmlchars with KEEP - and the exporter using layout parameteirzed using REMOVE
> 
> OK, the default preview of JabRef needs to be adapted maybe... The normal latex2unicode should be able to handle most cases, thus I think, there probably is no adaption necessary?

Your idea of an `enum BraceHandling` for `HTMLChars` is great, I love it. What I can do is modify `HTMLChars` to support both modes (i.e. remove braces or not), the default being `REMOVE`. I will only have to modify the preview to use the `KEEP` mode, which allows backwards compatibility in avoiding modifying all `*.layout` files.

**Comment by koppor** [comment]
> Your idea of an `enum BraceHandling` for `HTMLChars` is great, I love it. What I can do is modify `HTMLChars` to support both modes (i.e. remove braces or not), the default being `REMOVE`.

Add a constructor accepting the mode. The default constructor calls the other one with `REMOVE`.

Then add some ifs in the code covering the two cases. - The places should be easy to find as they are the places of your modifications.

Modifying the tests will be more hard, but required, too.

Note: I am aware this is not OO style at all. However, rewriting in OO-style would make the class even more complex and not help IMHO.

>  I will only have to modify the preview to use the `KEEP` mode, which allows backwards compatibility in avoiding modifying all `*.layout` files. WDYT ?

Yes!

File: src/main/java/org/jabref/logic/layout/format/HTMLChars.java [hunk 6]
```diff
@@ -23,6 +23,15 @@ public class HTMLChars implements LayoutFormatter {
      * */
     private static final Pattern HTML_ENTITY_PATTERN = Pattern.compile("&(?!(?:[a-z0-9]+|#[0-9]{1,6}|#x[0-9a-fA-F]{1,6});)");
 
+    private boolean keep = false;
+
+    @Override
+    public void setArgument(String arg) {
+        if ("keep".equals(arg)) {
+            this.keep = true;
+        }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/HTMLChars.java:32]: I would also use the full name `keepCurlyBraces` here, because otherwise, a reader does not understand what `keep` means. Future work could keep latex commands or something else.

```suggestion
        this.keep = "keepCurlyBraces".equalsIgnoreCase(arg);
```

File: src/main/java/org/jabref/logic/layout/format/HTMLChars.java [hunk 7]
```diff
@@ -23,6 +23,15 @@ public class HTMLChars implements LayoutFormatter {
      * */
     private static final Pattern HTML_ENTITY_PATTERN = Pattern.compile("&(?!(?:[a-z0-9]+|#[0-9]{1,6}|#x[0-9a-fA-F]{1,6});)");
 
+    private boolean keep = false;
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/HTMLChars.java:26]: Rename to `keepCurlyBraces`

File: src/main/java/org/jabref/migrations/PreferencesMigrations.java [hunk 8]
```diff
@@ -329,8 +329,8 @@ private static void migrateTypedKeyPrefs(JabRefPreferences prefs, Preferences ol
      */
     protected static void upgradePreviewStyle(JabRefPreferences prefs) {
         String currentPreviewStyle = prefs.get(JabRefPreferences.PREVIEW_STYLE);
-        String migratedStyle = currentPreviewStyle.replace("\\begin{review}<BR><BR><b>Review: </b> \\format[HTMLChars]{\\review} \\end{review}", "\\begin{comment}<BR><BR><b>Comment: </b> \\format[Markdown,HTMLChars]{\\comment} \\end{comment}")
-                                                  .replace("\\format[HTMLChars]{\\comment}", "\\format[Markdown,HTMLChars]{\\comment}")
+        String migratedStyle = currentPreviewStyle.replace("\\begin{review}<BR><BR><b>Review: </b> \\format[HTMLChars]{\\review} \\end{review}", "\\begin{comment}<BR><BR><b>Comment: </b> \\format[Markdown,HTMLChars(keep)]{\\comment} \\end{comment}")
+                                                  .replace("\\format[HTMLChars]{\\comment}", "\\format[Markdown,HTMLChars(keep)]{\\comment}")
```
Comment (koppor) [inline src/main/java/org/jabref/migrations/PreferencesMigrations.java:333]: Add a migration from 

    \\format[Markdown,HTMLChars]{\\comment}

to

    \\format[Markdown,HTMLChars(keepCurlyBraces)]{\\comment}

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java (hunk 3 above)
Comment (koppor) [reply src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:119]: Since the existing layouter is tested now, please add this back.

File: src/test/java/org/jabref/logic/layout/format/MarkdownFormatterTest.java [hunk 9]
```diff
@@ -21,6 +21,41 @@ void formatWhenFormattingPlainTextThenReturnsTextWrappedInParagraph() {
         assertEquals("<p>Hello World</p>", markdownFormatter.format("Hello World"));
     }
 
+    @Test
+    void formatWhenFormattingHeaderThenReturnsHeaderInHtml() {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/MarkdownFormatterTest.java:25]: You could convert all of them to ParameterizedTests!

All the tests seems to test some markdown-to-html functionality, which JabRef does not implement by itself, but through a library.

I would remove them and keep `formatWhenFormattingCodeBlockThenReturnsCodeBlockInHtml() {` only.

And then add some tests with empty lines to see how the converter deals with blocks. Are they also removed - or is only the code block the issue?

What about code blocks "recommended" by [MD031 of markdown-lint](https://github.com/DavidAnson/markdownlint/blob/main/doc/md031.md)?

````java
"""
   First line

   Second line

   ```java
   String test;
   ```
"""
````

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/test/java/org/jabref/logic/layout/format/MarkdownFormatterTest.java (hunk 9 above)
Comment (JEANPRBT) [reply src/test/java/org/jabref/logic/layout/format/MarkdownFormatterTest.java]: After having written some tests I get what's going on about new line formatting. The current formatter uses HTML `<p>` tags for each line. By default, HTML inserts a new line between each `<p>` tag, which is why in `MarkdownFormatter` we remove the `\n` before returning formatted strings, to avoid jumping two lines instead of one. But in `<pre>` tags which include the code in Mardown, there are no such `<p>` tags to indicate the new lines. When JabRef removes the `\n` before returning the final string, we lose track of those new lines in code blocks.

File: src/main/java/org/jabref/logic/layout/format/HTMLChars.java [hunk 10]
```diff
@@ -23,6 +23,15 @@ public class HTMLChars implements LayoutFormatter {
      * */
     private static final Pattern HTML_ENTITY_PATTERN = Pattern.compile("&(?!(?:[a-z0-9]+|#[0-9]{1,6}|#x[0-9a-fA-F]{1,6});)");
 
+    private boolean keepCurlyBraces = false;
+
+    @Override
+    public void setArgument(String arg) {
+        if ("keepCurlyBraces".equals(arg)) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/HTMLChars.java:30]: Please use equalsIgnoreCase. The CPU usage is neglectable in comparison to user experience.

File: src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java [hunk 11]
```diff
@@ -52,15 +57,16 @@ private static Stream<Arguments> provideLaTeXHighlightingFormattingData() {
         return Stream.of(
                 Arguments.of("<em>hallo</em>", "\\emph{hallo}"),
                 Arguments.of("<em>hallo</em>", "{\\emph hallo}"),
-                Arguments.of("<em>hallo</em>", "{\\em hallo}"),
```
Comment (koppor) [inline src/test/java/org/jabref/logic/layout/format/HTMLCharsTest.java:55]: No change in the existing formatter tests. Please only additions.

**Review by koppor — CHANGES_REQUESTED** [review]
